### PR TITLE
Afs ws authentication

### PR DESCRIPTION
**Pull request by geofjamg**

### CODE REVIEW

File: afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java [hunk 1]
```diff
@@ -24,21 +29,30 @@
...
-    public List<AppFileSystem> getFileSystems(ComputationManager computationManager) {
-        return configs.stream()
-                .map(config ->  {
-                    RemoteAppStorage storage = new RemoteAppStorage(config.getFileSystemName(), config.getUrl());
-                    RemoteListenableAppStorage listenableStorage = new RemoteListenableAppStorage(storage, config.getUrl());
-                    RemoteTaskMonitor taskMonitor = new RemoteTaskMonitor(config.getFileSystemName(), config.getUrl());
-                    return new AppFileSystem(config.getFileSystemName(), true, listenableStorage, taskMonitor);
-                })
-                .collect(Collectors.toList());
+    public List<AppFileSystem> getFileSystems(AppFileSystemProviderContext context) {
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java]: Assert that context is not null

File: afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java [hunk 2]
```diff
@@ -24,21 +29,30 @@
...
+        if (context.getToken() == null) {
+            return Collections.emptyList();
+        } else {
+            URI uri = config.get().getRestUri();
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java]: config.get() could return a null instance: you should check it

File: afs/afs-ws/afs-ws-server-utils/pom.xml [hunk 3]
```diff
@@ -36,13 +36,22 @@
...
         <!-- Test dependencies -->
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>powsybl-afs-ws-utils</artifactId>
+            <version>${project.version}</version>
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-server-utils/pom.xml]: Is scope test missing?

File: afs/afs-core/src/main/java/com/powsybl/afs/AppFileSystemProviderContext.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Objects;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class AppFileSystemProviderContext {
+
+    private final ComputationManager computationManager;
+
+    private String token;
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/AppFileSystemProviderContext.java]: Done

File: afs/afs-core/src/main/java/com/powsybl/afs/ServiceCreationContext.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.afs;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class ServiceCreationContext {
+
+    private String token;
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/ServiceCreationContext.java]: Done

File: afs/afs-local/src/main/java/com/powsybl/afs/local/LocalAppFileSystemProvider.java [hunk 6]
```diff
@@ -43,9 +43,9 @@ public LocalAppFileSystemProvider(List<LocalAppFileSystemConfig> configs, List<L
     }
 
     @Override
-    public List<AppFileSystem> getFileSystems(ComputationManager computationManager) {
+    public List<AppFileSystem> getFileSystems(AppFileSystemProviderContext context) {
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-local/src/main/java/com/powsybl/afs/local/LocalAppFileSystemProvider.java]: Done

File: afs/afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java [hunk 7]
```diff
@@ -42,10 +44,13 @@
     private static final String NODE_ID = "nodeId";
     private static final String NODE_PATH = "fileSystems/{fileSystemName}/nodes/{nodeId}";
 
-    private final Supplier<RemoteNetworkCacheServiceConfig> configSupplier;
+    private final Supplier<RemoteServiceConfig> configSupplier;
+
+    private String token;
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-network/afs-network-client/src/main/java/com/powsybl/afs/network/client/RemoteNetworkCacheService.java]: Done

File: afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/ClientUtils.java [hunk 8]
```diff
@@ -101,4 +108,34 @@ public static void checkOk(Response response) {
...
+        Objects.requireNonNull(baseUri);
+        Objects.requireNonNull(login);
+        Objects.requireNonNull(password);
+
+        Client client = ClientUtils.createClient()
+                .register(new JsonProvider());
+        try {
+            Form form = new Form()
+                    .param("login", login)
+                    .param("password", password);
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-ws/afs-ws-client-utils/src/main/java/com/powsybl/afs/ws/client/utils/ClientUtils.java]: Plain but using https connection

File: afs/afs-ws/afs-ws-client-utils/src/test/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfigTest.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class RemoteServiceConfigTest {
+
+    @Test
+    public void test() {
+        RemoteServiceConfig config = new RemoteServiceConfig("host", "test", 443, true);
+        assertEquals("https://host:443/test", config.getRestUri().toString());
+        assertEquals("wss://host:443/test", config.getWsUri().toString());
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply afs/afs-ws/afs-ws-client-utils/src/test/java/com/powsybl/afs/ws/client/utils/RemoteServiceConfigTest.java]: Done

File: afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java (hunk 1 above)
Comment (geofjamg) [reply afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java]: Done

File: afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java (hunk 2 above)
Comment (geofjamg) [reply afs/afs-ws/afs-ws-client/src/main/java/com/powsybl/afs/ws/client/RemoteAppFileSystemProvider.java]: Done

File: afs/afs-ws/afs-ws-server-utils/pom.xml (hunk 3 above)
Comment (geofjamg) [reply afs/afs-ws/afs-ws-server-utils/pom.xml]: No it just in the wrong section, moved to compile dependencies

File: afs/afs-core/src/main/java/com/powsybl/afs/AppData.java [hunk 10]
```diff
@@ -137,15 +162,29 @@ public void addFileSystem(AppFileSystem fileSystem) {
...
         return fileSystems.get(name);
     }
 
+    public void setTokenProvider(SecurityTokenProvider tokenProvider) {
+        this.tokenProvider = Objects.requireNonNull(tokenProvider);
+        // clean loaded file systems and services
+        if (fileSystems != null) {
+            for (AppFileSystem fileSystem : fileSystems.values()) {
+                fileSystem.close();
+            }
```
Comment (mathbagu) [inline afs/afs-core/src/main/java/com/powsybl/afs/AppData.java]: Simply call the "close" method?

File: afs/afs-ws/afs-ws-server-utils/pom.xml [hunk 11]
```diff
@@ -29,13 +29,22 @@
...
 
         <!-- Provided dependencies -->
         <dependency>
             <groupId>javax</groupId>
             <artifactId>javaee-api</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt</artifactId>
```
Comment (mathbagu) [inline afs/afs-ws/afs-ws-server-utils/pom.xml]: Wrong section or scope is missing?

File: afs/afs-core/src/main/java/com/powsybl/afs/AppData.java (hunk 10 above)
Comment (geofjamg) [reply afs/afs-core/src/main/java/com/powsybl/afs/AppData.java]: Moved to closeFileSystems()

File: afs/afs-ws/afs-ws-server-utils/pom.xml (hunk 11 above)
Comment (geofjamg) [reply afs/afs-ws/afs-ws-server-utils/pom.xml]: Wrong section, fixed

**Pull request merged**